source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -868,14 +883,13 @@ bssl::UniquePtr<SSL> ClientContextImpl::newSsl(const Network::TransportSocketOpt
     SSL_set_verify(ssl_con.get(), SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, nullptr);
   }
 
-  if (options && !options->applicationProtocolListOverride().empty()) {
-    std::vector<uint8_t> parsed_override_alpn =
-        parseAlpnProtocols(absl::StrJoin(options->applicationProtocolListOverride(), ","));
-    if (!parsed_override_alpn.empty()) {
-      const int rc = SSL_set_alpn_protos(ssl_con.get(), parsed_override_alpn.data(),
-                                         parsed_override_alpn.size());
-      RELEASE_ASSERT(rc == 0, Utility::getLastCryptoError().value_or(""));
-    }
+  bool has_alpn_defined = !parsed_alpn_protocols_.empty();
+  if (options) {
+    has_alpn_defined |= parseAndSetAlpn(options->applicationProtocolListOverride(), *ssl_con);
+  }
+
+  if (options && !has_alpn_defined) {
+    parseAndSetAlpn(options->applicationProtocolListFallback(), *ssl_con);
```

Comment:
OK I think this makes sense now. Can you add more comments though about precedence? It takes a bit of eye squinting to figure out what is going on here.